File: raytrace/Object/Shape/BezierPatch.cpp
#include "Object/Shape/BezierPatch.hpp"

namespace Object {
	namespace Shape {
		BezierPatch::BezierPatch(int width, int height, std::vector<Math::Point> &&controlPoints)
		{
			std::vector<Math::Point> points;
			std::vector<Math::Normal> normals;
			std::vector<Math::Vector> du;
			std::vector<Math::Vector> dv;

			for (int j = 0; j < height; j++) {
				for (int i = 0; i < width; i++) {
					float s = float(i) / (width - 1);
					float t = float(j) / (height - 1);
					float s1 = float((i < width / 2) ? (i + 1) : (i - 1)) / (width - 1);
					float t1 = float((j < height / 2) ? (j + 1) : (j - 1)) / (height - 1);

					float Bs[4] = { (1 - s)*(1 - s)*(1 - s), 3 * s*(1 - s)*(1 - s), 3 * s*s*(1 - s), s*s*s };
					float Bt[4] = { (1 - t)*(1 - t)*(1 - t), 3 * t*(1 - t)*(1 - t), 3 * t*t*(1 - t), t*t*t };
					float Bds[4] = { -3 * (1 - s)*(1 - s), 3 * (1 - s)*(1 - s) - 3 * 2 * s * (1 - s), 3 * 2 * s*(1 - s) - 3 * s * s, 3 * s*s };
					float Bdt[4] = { -3 * (1 - t)*(1 - t), 3 * (1 - t)*(1 - t) - 3 * 2 * t * (1 - t), 3 * 2 * t*(1 - t) - 3 * t * t, 3 * t*t };

					float Bs1[4] = { (1 - s1)*(1 - s1)*(1 - s1), 3 * s1*(1 - s1)*(1 - s1), 3 * s1*s1*(1 - s1), s1*s1*s1 };
					float Bt1[4] = { (1 - t1)*(1 - t1)*(1 - t1), 3 * t1*(1 - t1)*(1 - t1), 3 * t1*t1*(1 - t1), t1*t1*t1 };
					float Bds1[4] = { -3 * (1 - s1)*(1 - s1), 3 * (1 - s1)*(1 - s1) - 3 * 2 * s1 * (1 - s1), 3 * 2 * s1*(1 - s1) - 3 * s1 * s1, 3 * s1*s1 };
					float Bdt1[4] = { -3 * (1 - t1)*(1 - t1), 3 * (1 - t1)*(1 - t1) - 3 * 2 * t1 * (1 - t1), 3 * 2 * t1*(1 - t1) - 3 * t1 * t1, 3 * t1*t1 };

					Math::Vector p;
					Math::Vector ds;
					Math::Vector dt;
					Math::Vector ds1;
					Math::Vector dt1;
					for (int k = 0; k < 4; k++) {
						for (int l = 0; l < 4; l++) {
							p = p + Math::Vector(controlPoints[k * 4 + l]) * (Bs[l] * Bt[k]);
							ds = ds + Math::Vector(controlPoints[k * 4 + l]) * (Bds[l] * Bt[k]);
							dt = dt + Math::Vector(controlPoints[k * 4 + l]) * (Bs[l] * Bdt[k]);
							ds1 = ds1 + Math::Vector(controlPoints[k * 4 + l]) * (Bds1[l] * Bt1[k]);
							dt1 = dt1 + Math::Vector(controlPoints[k * 4 + l]) * (Bs1[l] * Bdt1[k]);
						}
					}

					if (ds.magnitude2() < 0.0000001) {
						ds = ds1;
					}
					if (dt.magnitude2() < 0.0000001) {
						dt = dt1;
					}

					points.push_back(Math::Point(p));
					du.push_back(ds.normalize());
					dv.push_back(dt.normalize());
					normals.push_back(Math::Normal(ds % dt).normalize());
				}
			}

			mGrid = std::make_unique<Grid>(width, height, std::move(points), std::move(normals), std::move(du), std::move(dv));
		}

		bool BezierPatch::intersect(const Math::Ray &ray, Intersection &intersection) const
		{
			return mGrid->intersect(ray, intersection);
		}

		BoundingVolume BezierPatch::boundingVolume(const Math::Transformation &transformation) const
		{
			return mGrid->boundingVolume(transformation);
		}
	}
}

File: src/object/shape/bezier_patch.rs
use crate::geo;
use crate::object;

use geo::Bivec3;
use geo::Normal3;
use geo::Point3;
use geo::Vec3;

use object::Shape;
use object::ShapeIntersection;
use object::shape::Grid;
use object::shape::GridVertex;

pub struct BezierPatch {
    grid : Grid
}

impl BezierPatch {
    pub fn new(width : usize, height : usize, control_points : [Point3; 16]) -> BezierPatch {
        let mut vertices = Vec::new();

        for j in 0..height {
            for i in 0..width {
                let s = (i as f32) / ((width - 1) as f32);
                let t = (j as f32) / ((height - 1) as f32);
                let s1 = ((if i < width / 2 { i + 1 } else { i - 1 }) as f32) / ((width - 1) as f32);
                let t1 = ((if j < height / 2 { j + 1 } else { j - 1 }) as f32) / ((height - 1) as f32);
          
                let bs = [(1.0 - s).powi(3), 3.0 * s * (1.0 - s).powi(2), 3.0 * s.powi(2) * (1.0 - s), s.powi(3)];
                let bt = [(1.0 - t).powi(3), 3.0 * t * (1.0 - t).powi(2), 3.0 * t.powi(2) * (1.0 - t), t.powi(3)];
                let bds = [-3.0 * (1.0 - s).powi(2), 3.0 * (1.0 - s).powi(2) - 3.0 * 2.0 * s * (1.0 - s), 3.0 * 2.0 * s * (1.0 - s) - 3.0 * s.powi(2), 3.0 * s.powi(2)];
                let bdt = [-3.0 * (1.0 - t).powi(2), 3.0 * (1.0 - t).powi(2) - 3.0 * 2.0 * t * (1.0 - t), 3.0 * 2.0 * t * (1.0 - t) - 3.0 * t.powi(2), 3.0 * t.powi(2)];

                let bs1 = [(1.0 - s1).powi(3), 3.0 * s1 * (1.0 - s1).powi(2), 3.0 * s1.powi(2) * (1.0 - s1), s1.powi(3)];
                let bt1 = [(1.0 - t1).powi(3), 3.0 * t1 * (1.0 - t1).powi(2), 3.0 * t1.powi(2) * (1.0 - t1), t1.powi(3)];
                let bds1 = [-3.0 * (1.0 - s1).powi(2), 3.0 * (1.0 - s1).powi(2) - 3.0 * 2.0 * s1 * (1.0 - s1), 3.0 * 2.0 * s1 * (1.0 - s1) - 3.0 * s1.powi(2), 3.0 * s1.powi(2)];
                let bdt1 = [-3.0 * (1.0 - t1).powi(2), 3.0 * (1.0 - t1).powi(2) - 3.0 * 2.0 * t1 * (1.0 - t1), 3.0 * 2.0 * t1 * (1.0 - t1) - 3.0 * t1.powi(2), 3.0 * t1.powi(2)];

                let mut p = Vec3::ZERO;
                let mut ds = Vec3::ZERO;
                let mut dt = Vec3::ZERO;
                let mut ds1 = Vec3::ZERO;
                let mut dt1 = Vec3::ZERO;

                for k in 0..4 {
                    for l in 0..4 {
                        let c = control_points[k*4 + l].to_vec3();

                        p += c * (bs[l] * bt[k]);
                        ds += c * (bds[l] * bt[k]);
                        dt += c * (bs[l] * bdt[k]);
                        ds1 += c * bds1[l] * bt1[k];
                        dt1 += c * (bs1[l] * bdt1[k]);
                    }
                }

                if ds.mag2() < 0.0000001 {
                    ds = ds1;
                }

                if dt.mag2() < 0.0000001 {
                    dt = dt1;
                }

                let vertex = GridVertex{point: Point3::from_vec3(p), tangent: Bivec3::new(ds, dt), normal: Normal3::from_vec3((ds % dt).normalize())};
                vertices.push(vertex);
            }
        }
        
        let grid = Grid::new(width, height, vertices);
        return BezierPatch {grid};
    }
}

impl Shape for BezierPatch {
    fn bounding_volume(&self, xform : geo::Transformation) -> object::BoundingVolume {
        return self.grid.bounding_volume(xform);
    }

    fn intersect(&self, ray : geo::Ray, max_distance : f32, closest : bool) -> Option<ShapeIntersection> {
        return self.grid.intersect(ray, max_distance, closest);
    }
}